core-primitives/utils/src/lib.rs

```diff
@@ -0,0 +1,92 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+//! General utility functions.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(all(feature = "std", feature = "sgx"))]
+compile_error!("feature \"std\" and feature \"sgx\" cannot be enabled at the same time");
+
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+extern crate sgx_tstd as std;
+
+// re-export module to properly feature gate sgx and regular std environment
+#[cfg(all(not(feature = "std"), feature = "sgx"))]
+pub mod sgx_reexport_prelude {
+	pub use thiserror_sgx as thiserror;
+}
+
+mod error;
+
+pub use error::{Error, Result};
+use frame_support::ensure;
+use std::{string::String, vec::Vec};
+
+/// Hex encodes given data and preappends a "0x".
+pub fn hex_encode(data: Vec<u8>) -> String {
+	let mut hex_str = hex::encode(data);
+	hex_str.insert_str(0, "0x");
+	hex_str
+}
```

Comment:
I think we can become more ergonomic with:

```rust
pub trait ToHexPrefixed {
   fn toHex(&self) -> String;
}

impl<T: Encode> ToHexPrefixed for T {
    fn toHex(&self) -> String {
        hex_encode(self.encode); 
    }
}
```

and a similar pattern for decode with the following function signature:
```rust

pub trait FromHexPrefixed {
   Self::Output;

   fn fromHex(msg: &str) -> Result<Self::Output, SomeError>;
}

impl<T: Decode> FromHexPrefixed for T {
    Self::Output = T;


    fn fromHex(msg: &[u8]) -> Result<Self::Output, SomeError> {
        ...
        let unhex =  decode_hex(msg)?;
        Decode::decode(unhex).map_err()?;
   }
}

```

btw. I am pretty sure the `hex::encode` and `hex::decode` can take references.

Then we can use it like:

```rust

let hex_thing = my_thing.to_hex();
let my_same_thing = MyThing::from_hex(&hex_thing).unwrap();
```

Reply:
Not 100% that the decode pattern works like this, there might be some rust technical issues with that.

Reply:
I like that suggestion. Let me try it